### PR TITLE
Small index optimisation

### PR DESCRIPTION
**Pull request by fommil**
@rorygraves you're it. Please merge if you're happy and it goes green.

### CODE REVIEW

**Comment by fommil** [comment]
@jschlather now you get me back, can you have a quick review please incase you spot anything obvious?

File: core/src/main/scala/org/ensime/indexer/SearchService.scala [hunk 1]
```diff
@@ -117,11 +118,15 @@ class SearchService(
     // index all the given bases and return number of rows written
     def indexBases(bases: Set[FileObject], checks: Seq[FileCheck]): Future[Int] = {
       log.info("Indexing bases...")
-      val checksLookup: Map[String, FileCheck] = checks.map(check => (check.filename -> check)).toMap
-      val basesWithChecks: Set[(FileObject, Option[FileCheck])] = bases.map { base =>
-        (base, checksLookup.get(base.getName().getURI()))
+      val known: Map[FileObject, FileCheck] = checks.flatMap { check =>
+        val f = check.file
+        if (bases(f)) Some(check.file -> check)
```
Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/SearchService.scala]: assumes that `FileObject` correctly implements `equals`, of course.

Comment (jschlather) [inline core/src/main/scala/org/ensime/indexer/SearchService.scala]: It's not clear to me that FileObject does implement equals properly. Looking at the source for [AbstractFileObject](http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-vfs2/2.0/org/apache/commons/vfs2/provider/AbstractFileObject.java#AbstractFileObject) and [ZipFileObject](http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-vfs2/2.0/org/apache/commons/vfs2/provider/zip/ZipFileObject.java?av=f) it doesn't seem like it's implemented. If it inherits from `Object`, is the map always going to be empty?

**Comment by jschlather** [comment]
It looks fine to me as long as the set lookup works properly.

Also FWIW, I did some profiling of SearchService startup to see what was taking the longest. As a disclaimer, I only looked at how long the SearchServiceSpec tests took, so it might be beneficial to look at a large project. But, the majority of the time spent seemed to spent in the `ClassfileIndexer` calling `indexClassFile`. I didn't look too deeply at that classfile, but there might be some more performance gains in there.

**Comment by fommil** [comment]
OK, let's hold until we can confirm if equals is implemented correctly.

**Comment by fommil** [comment]
you're quite right, `.equals` is not implemented https://github.com/apache/commons-vfs/blob/VFS-2.0/core/src/main/java/org/apache/commons/vfs/provider/local/LocalFile.java

The VFS library just gets worse and worse. It's such a massive behemoth and doesn't seem to do really basic stuff. Classic Apache library.

**Comment by fommil** [comment]
actually, I'm just closing this PR... it's not on my top priority list. As our resident performance optimisation guru, I leave this to @rorygraves to pick up if he'd like to encourage further use of `breakOut` vs `.toMap` and friends.

**Comment by fommil** [comment]
@jschlather yeah one of the easiest performance boosts we could get would be to rewrite the `DescriptorParser` in parboiled2 instead of parboiled.

**Comment by jschlather** [comment]
Interesting, I might take a crack at that when I have some free time and see how much work it would be. If no one else is working on it.

**Comment by rorygraves** [comment]
Ok, did some quick profiler of the indexer startup - its basically all parboiled

**Comment by aemoncannon** [comment]
:-)

**Comment by jschlather** [comment]
Yeah, that's what it was looking like for me too. @rorygraves  Out of curiosity, what did you use to profile the indexer startup?

**Comment by rorygraves** [comment]
Yourkit - They very nicely provided a couple of OSS licenses for ensime

**Comment by fommil** [comment]
@rorygraves run that on windows and see a very different picture ;-) all h2/slick.

But parboiled2 is definitely going to help a **lot** on all platforms.

**Pull request closed without merging**